### PR TITLE
Add GitHubVulnerabilities2Db Job

### PR DESCRIPTION
**Pull request by scottbommarito**
Part of https://github.com/nuget/nugetgallery/issues/7648

This PR adds the GitHubVulnerabilities2Db job itself, which primarily consists of querying GitHub's GraphQL API and then converting the response into our DB entities.

### CODE REVIEW

File: src/GitHubVulnerabilities2Db/Collector/AdvisoryCollector.cs [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+            _cursor = cursor ?? throw new ArgumentNullException(nameof(cursor));
+            _queryService = queryService ?? throw new ArgumentNullException(nameof(queryService));
+            _ingestor = ingestor ?? throw new ArgumentNullException(nameof(ingestor));
+        }
+
+        private readonly ReadWriteCursor<DateTimeOffset> _cursor;
+        private readonly IAdvisoryCollectorQueryService _queryService;
+        private readonly IAdvisoryIngestor _ingestor;
+
+        public async Task<bool> Process(CancellationToken token)
```
Comment (loic-sharma) [inline src/GitHubVulnerabilities2Db/Collector/AdvisoryCollector.cs]:
```suggestion
        public async Task<bool> ProcessAsync(CancellationToken token)
```

File: src/GitHubVulnerabilities2Db/Collector/AdvisoryCollector.cs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+        {
+            var advisories = await _queryService.GetAdvisoriesSinceAsync(_cursor, token);
+            var hasAdvisories = advisories != null && advisories.Any();
+            if (hasAdvisories)
```
Comment (loic-sharma) [inline src/GitHubVulnerabilities2Db/Collector/AdvisoryCollector.cs]: Please add logging

File: src/GitHubVulnerabilities2Db/GraphQL/Node.cs [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    /// </summary>
+    public interface INode
+    {
+        DateTimeOffset UpdatedAt { get; set; }
+    }
+
+    /// <summary>
+    /// https://developer.github.com/v4/object/securityadvisory/
+    /// </summary>
+    public class SecurityAdvisory : INode
```
Comment (loic-sharma) [inline src/GitHubVulnerabilities2Db/GraphQL/Node.cs]: Maybe rename this file to something like `SecurityAdvisory.cs`?

Comment (loic-sharma) [reply src/GitHubVulnerabilities2Db/GraphQL/Node.cs]: Also, maybe move `INode` to `QueryResponse.cs`?

Comment (scottbommarito) [reply src/GitHubVulnerabilities2Db/GraphQL/Node.cs]: Hmm...the reason why I grouped these together was
- `SecurityAdvisory` and `SecurityVulnerability` are both `INode`s
- the other classes are only used by them
- I wanted to add additional nodes to this file

But I think your split, by moving `INode` into `QueryResponse` and keeping this file for just the security entities makes sense.

File: src/GitHubVulnerabilities2Db/GraphQL/QueryService.cs [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+            var queryJObject = new JObject
+            {
+                ["query"] = query
+            };
+
+            var response = await GetResponseFromQuery(queryJObject.ToString(), token);
+            return JsonConvert.DeserializeObject<QueryResponse>(response);
+        }
+
+        private async Task<string> GetResponseFromQuery(string query, CancellationToken token)
```
Comment (loic-sharma) [inline src/GitHubVulnerabilities2Db/GraphQL/QueryService.cs]:
```suggestion
        private async Task<string> GetResponseFromQueryAsync(string query, CancellationToken token)
```

Also, what do you think of the name `ExecuteQueryAsync`?

File: src/GitHubVulnerabilities2Db/GraphQL/QueryService.cs [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+        public QueryService(
+            InitializationConfiguration initializationConfiguration,
+            HttpClient client)
+        {
+            InitializationConfiguration = initializationConfiguration ?? throw new ArgumentNullException(nameof(initializationConfiguration));
+            Client = client ?? throw new ArgumentNullException(nameof(client));
+        }
+
+        public InitializationConfiguration InitializationConfiguration { get; set; }
+        public HttpClient Client { get; set; }
```
Comment (loic-sharma) [inline src/GitHubVulnerabilities2Db/GraphQL/QueryService.cs]: Do these need to be public? And settable? Can we make these readonly fields?

File: src/GitHubVulnerabilities2Db/GraphQL/QueryService.cs (hunk 4 above)
Comment (scottbommarito) [reply src/GitHubVulnerabilities2Db/GraphQL/QueryService.cs]: This method is for specifically making the web request to the Graph QL API, I don't think `ExecuteQueryAsync` is appropriate here. How do you feel about `MakeWebRequestAsync`?

File: src/GitHubVulnerabilities2Db/Configuration/InitializationConfiguration.cs [hunk 6]
```diff
@@ -0,0 +1,35 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+
+namespace GitHubVulnerabilities2Db.Configuration
+{
+    public class InitializationConfiguration
```
Comment (loic-sharma) [inline src/GitHubVulnerabilities2Db/Configuration/InitializationConfiguration.cs]: This config name is a tad specific, especially since it is used by things like `QueryService`. Maybe call this something like `GitHubVulnerabilitiesConfiguration`?

Comment (scottbommarito) [reply src/GitHubVulnerabilities2Db/Configuration/InitializationConfiguration.cs]: We have a pattern of naming the config for jobs `InitializationConfiguration`. I was initially considering naming it what you suggested but stuck with the existing pattern instead. Note that some jobs do not follow this pattern, but most do.

File: src/GitHubVulnerabilities2Db/Collector/AdvisoryCollector.cs (hunk 2 above)
Comment (scottbommarito) [reply src/GitHubVulnerabilities2Db/Collector/AdvisoryCollector.cs]: Added logging. Not sure what telemetry you think we should add--we can query the logs like telemetry anyway.

File: src/GitHubVulnerabilities2Db/GraphQL/QueryService.cs (hunk 4 above)
Comment (loic-sharma) [reply src/GitHubVulnerabilities2Db/GraphQL/QueryService.cs]: I don't feel strongly about this, all names are fine by me 😊

File: src/GitHubVulnerabilities2Db/Configuration/InitializationConfiguration.cs (hunk 6 above)
Comment (loic-sharma) [reply src/GitHubVulnerabilities2Db/Configuration/InitializationConfiguration.cs]: Mmmmm, this is a strange pattern

File: src/GitHubVulnerabilities2Db/Collector/IAdvisoryCollectorQueryBuilder.cs [hunk 7]
```diff
@@ -0,0 +1,15 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using GitHubVulnerabilities2Db.GraphQL;
+
+namespace GitHubVulnerabilities2Db.Collector
+{
+    public interface IAdvisoryCollectorQueryBuilder
```
Comment (loic-sharma) [inline src/GitHubVulnerabilities2Db/Collector/IAdvisoryCollectorQueryBuilder.cs]: Thoughts on calling this `IAdvisoryQueryBuilder`? "Collector" doesn't really add much here, it's in the namespace.

File: src/GitHubVulnerabilities2Db/Collector/IAdvisoryCollectorQueryService.cs [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+using System.Threading.Tasks;
+using GitHubVulnerabilities2Db.GraphQL;
+using NuGet.Services.Cursor;
+
+namespace GitHubVulnerabilities2Db.Collector
+{
+    /// <summary>
+    /// Wrapper around <see cref="IQueryService"/> to make it easier to query for <see cref="SecurityAdvisory"/>s using a cursor.
+    /// </summary>
+    public interface IAdvisoryCollectorQueryService
```
Comment (loic-sharma) [inline src/GitHubVulnerabilities2Db/Collector/IAdvisoryCollectorQueryService.cs]: Thoughts on calling this `IAdvisoryQueryService`? "Collector" doesn't really add much here, it's in the namespace.

File: src/GitHubVulnerabilities2Db/Configuration/InitializationConfiguration.cs (hunk 6 above)
Comment (scottbommarito) [reply src/GitHubVulnerabilities2Db/Configuration/InitializationConfiguration.cs]: I think it would be better if I got rid of it, so I will go forward with what you recommend.

File: src/GitHubVulnerabilities2Db/Collector/IAdvisoryCollectorQueryBuilder.cs (hunk 7 above)
Comment (scottbommarito) [reply src/GitHubVulnerabilities2Db/Collector/IAdvisoryCollectorQueryBuilder.cs]: Sounds good!

File: src/GitHubVulnerabilities2Db/Collector/IAdvisoryCollectorQueryService.cs (hunk 8 above)
Comment (scottbommarito) [reply src/GitHubVulnerabilities2Db/Collector/IAdvisoryCollectorQueryService.cs]: Sounds good!

File: src/GitHubVulnerabilities2Db/GraphQL/QueryService.cs [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+        public QueryService(
+            InitializationConfiguration initializationConfiguration,
+            HttpClient client)
+        {
+            _initializationConfiguration = initializationConfiguration ?? throw new ArgumentNullException(nameof(initializationConfiguration));
+            _client = client ?? throw new ArgumentNullException(nameof(client));
+        }
+
+        private readonly InitializationConfiguration _initializationConfiguration;
+        private readonly HttpClient _client;
```
Comment (loic-sharma) [inline src/GitHubVulnerabilities2Db/GraphQL/QueryService.cs]: Please move fields to before the constructor

File: src/GitHubVulnerabilities2Db/GraphQL/QueryService.cs [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+            };
+
+            var response = await MakeWebRequestAsync(queryJObject.ToString(), token);
+            return JsonConvert.DeserializeObject<QueryResponse>(response);
+        }
+
+        private async Task<string> MakeWebRequestAsync(string query, CancellationToken token)
+        {
+            using (var request = CreateRequest(query))
+            using (var response = await _client.SendAsync(request, HttpCompletionOption.ResponseContentRead, token))
```
Comment (loic-sharma) [inline src/GitHubVulnerabilities2Db/GraphQL/QueryService.cs]: `HttpCompletionOption.ResponseContentRead` is the default, consider skipping that parameter:

```suggestion
            using (var response = await _client.SendAsync(request, token))
```

Comment (scottbommarito) [reply src/GitHubVulnerabilities2Db/GraphQL/QueryService.cs]: I believe the reason I had that there was because I didn't see there was a method with both a `HttpRequestMessage` and a `CancellationToken`, but I see there is one!

File: src/GitHubVulnerabilities2Db/GraphQL/QueryService.cs (hunk 9 above)
Comment (scottbommarito) [reply src/GitHubVulnerabilities2Db/GraphQL/QueryService.cs]: Done! Tried to do this for all the classes. Let me know if you see one with it still wrong.

File: src/GitHubVulnerabilities2Db/Ingest/AdvisoryIngestor.cs [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+                await _packageVulnerabilityService.UpdateVulnerabilityAsync(vulnerability, wasWithdrawn);
+            }
+        }
+
+        private Tuple<PackageVulnerability, bool> FromAdvisory(SecurityAdvisory advisory)
+        {
+            var vulnerability = new PackageVulnerability
+            {
+                GitHubDatabaseKey = advisory.DatabaseId,
+                Severity = (PackageVulnerabilitySeverity)Enum.Parse(typeof(PackageVulnerabilitySeverity), advisory.Severity, ignoreCase: true),
```
Comment (loic-sharma) [inline src/GitHubVulnerabilities2Db/Ingest/AdvisoryIngestor.cs]: What if GitHub adds a new severity? I would be defensive about mapping a `SecurityAdvisory` to a `PackageVulnerability` and add thorough logging to make these kinds of issues easily debuggable.

File: tests/GitHubVulnerabilities2Db.Facts/GitHubVersionRangeParserFacts.cs [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+        [InlineData("a a b b c c")] // cannot have more than 3 parts
+        [InlineData("asdfasdf")] // must have 2 * n number of parts
+        [InlineData("<> 1.0.0")] // invalid symbol
+        [InlineData("> 1.0.0, > 2.0.0")] // min already specified
+        [InlineData("= 1.0.0, > 2.0.0")] // min already specified
+        [InlineData("> 1.0.0, = 2.0.0")] // min already specified
+        [InlineData("< 1.0.0, < 2.0.0")] // max already specified
+        [InlineData("= 1.0.0, < 2.0.0")] // max already specified
+        [InlineData("< 1.0.0, = 2.0.0")] // max already specified
+        [InlineData("= 1.0.0, = 2.0.0")] // min and max already specified
```
Comment (loic-sharma) [inline tests/GitHubVulnerabilities2Db.Facts/GitHubVersionRangeParserFacts.cs]: What's the expected result of `= 1.0.0, = 1.0.0`? It's a bit of a silly case 😊

File: src/GitHubVulnerabilities2Db/Ingest/AdvisoryIngestor.cs (hunk 11 above)
Comment (scottbommarito) [reply src/GitHubVulnerabilities2Db/Ingest/AdvisoryIngestor.cs]: Are you suggesting that the job should get stuck and be broken until we fix the severity mismatch? Or should I add an `Unknown` severity for this case?

File: src/GitHubVulnerabilities2Db/Job.cs [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+    public class Job : JsonConfigurationJob, IDisposable
+    {
+        private readonly HttpClient _client = new HttpClient();
+
+        public override async Task Run()
+        {
+            var collector = _serviceProvider.GetRequiredService<IAdvisoryCollector>();
+            using (var tokenSource = new CancellationTokenSource())
+            {
+                while (await collector.ProcessAsync(tokenSource.Token)) ;
```
Comment (loic-sharma) [inline src/GitHubVulnerabilities2Db/Job.cs]: Can you use `CancellationToken.None`?

File: tests/GitHubVulnerabilities2Db.Facts/GitHubVersionRangeParserFacts.cs (hunk 12 above)
Comment (scottbommarito) [reply tests/GitHubVulnerabilities2Db.Facts/GitHubVersionRangeParserFacts.cs]: Current implementation would throw as well, because I don't check if the min or max is the same, I just check if it's already set.

File: src/GitHubVulnerabilities2Db/Ingest/AdvisoryIngestor.cs (hunk 11 above)
Comment (loic-sharma) [reply src/GitHubVulnerabilities2Db/Ingest/AdvisoryIngestor.cs]: I think the job should be broken until we fix it, but, the logs should make it clear what's broken and why. Thoughts?

File: tests/GitHubVulnerabilities2Db.Facts/GitHubVersionRangeParserFacts.cs (hunk 12 above)
Comment (loic-sharma) [reply tests/GitHubVulnerabilities2Db.Facts/GitHubVersionRangeParserFacts.cs]: Do you think it's worth covering that corner case with a test?

File: src/GitHubVulnerabilities2Db/Ingest/AdvisoryIngestor.cs (hunk 11 above)
Comment (scottbommarito) [reply src/GitHubVulnerabilities2Db/Ingest/AdvisoryIngestor.cs]: Discussed this with @skofman1 offline, we both agreed with your idea. Looks like `Enum.Parse` will already throw if the value isn't a defined constant, so we're good here.

File: tests/GitHubVulnerabilities2Db.Facts/GitHubVersionRangeParserFacts.cs (hunk 12 above)
Comment (scottbommarito) [reply tests/GitHubVulnerabilities2Db.Facts/GitHubVersionRangeParserFacts.cs]: Sure haha.

File: src/GitHubVulnerabilities2Db/Job.cs (hunk 13 above)
Comment (scottbommarito) [reply src/GitHubVulnerabilities2Db/Job.cs]: Am using it now!

Maybe one day, `RunAsync` will pass in a token itself.

File: src/GitHubVulnerabilities2Db/GraphQL/QueryResponse.cs [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+
+using System;
+using System.Collections.Generic;
+
+namespace GitHubVulnerabilities2Db.GraphQL
+{
+    /// <summary>
+    /// A GraphQL query response object.
+    /// </summary>
+    public class QueryResponse
```
Comment (xavierdecoster) [inline src/GitHubVulnerabilities2Db/GraphQL/QueryResponse.cs]: nit: prefer one type per file

File: src/GitHubVulnerabilities2Db/GraphQL/QueryService.cs [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+using GitHubVulnerabilities2Db.Configuration;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+namespace GitHubVulnerabilities2Db.GraphQL
+{
+    public class QueryService : IQueryService
+    {
+        /// <remarks>
+        /// GitHub requires that every request includes a UserAgent.
```
Comment (xavierdecoster) [inline src/GitHubVulnerabilities2Db/GraphQL/QueryService.cs]: Sounds like a test assertion.

File: src/GitHubVulnerabilities2Db/Properties/AssemblyInfo.cs [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+// General Information about an assembly is controlled through the following
+// set of attributes. Change these attribute values to modify the information
+// associated with an assembly.
+[assembly: AssemblyTitle("GitHubVulnerabilities2Db")]
+[assembly: AssemblyDescription("")]
+[assembly: AssemblyConfiguration("")]
+[assembly: AssemblyCompany("")]
```
Comment (xavierdecoster) [inline src/GitHubVulnerabilities2Db/Properties/AssemblyInfo.cs]:
```suggestion
[assembly: AssemblyCompany(".NET Foundation")]
```

File: src/GitHubVulnerabilities2Db/Properties/AssemblyInfo.cs [hunk 17]
```diff
@@ -0,0 +1,36 @@
...
+[assembly: AssemblyProduct("GitHubVulnerabilities2Db")]
+[assembly: AssemblyCopyright("Copyright ©  2019")]
```
Comment (xavierdecoster) [inline src/GitHubVulnerabilities2Db/Properties/AssemblyInfo.cs]:
```suggestion
[assembly: AssemblyCopyright("\x00a9 .NET Foundation. All rights reserved.")]
```

File: src/GitHubVulnerabilities2Db/Properties/AssemblyInfo.cs [hunk 18]
```diff
@@ -0,0 +1,36 @@
+﻿using System.Reflection;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+// General Information about an assembly is controlled through the following
+// set of attributes. Change these attribute values to modify the information
+// associated with an assembly.
+[assembly: AssemblyTitle("GitHubVulnerabilities2Db")]
+[assembly: AssemblyDescription("")]
+[assembly: AssemblyConfiguration("")]
```
Comment (xavierdecoster) [inline src/GitHubVulnerabilities2Db/Properties/AssemblyInfo.cs]:
```suggestion
#if DEBUG
[assembly: AssemblyConfiguration("Debug")]
#else
[assembly: AssemblyConfiguration("Release")]
#endif
```

File: src/GitHubVulnerabilities2Db/Properties/AssemblyInfo.cs [hunk 19]
```diff
@@ -0,0 +1,36 @@
...
+//      Major Version
+//      Minor Version
+//      Build Number
+//      Revision
+//
+// You can specify all the values or you can default the Build and Revision Numbers
+// by using the '*' as shown below:
+// [assembly: AssemblyVersion("1.0.*")]
+[assembly: AssemblyVersion("1.0.0.0")]
+[assembly: AssemblyFileVersion("1.0.0.0")]
```
Comment (xavierdecoster) [inline src/GitHubVulnerabilities2Db/Properties/AssemblyInfo.cs]: Think you can remove these lines. Won't the build automatically inject the version attributes?

Comment (xavierdecoster) [inline src/GitHubVulnerabilities2Db/Properties/AssemblyInfo.cs]: Other projects in this solution have the following:

```suggestion
// The build will automatically inject the following attributes:
// AssemblyVersion, AssemblyFileVersion, AssemblyInformationalVersion, AssemblyMetadata (for Branch, CommitId, and BuildDateUtc)

[assembly: AssemblyMetadata("RepositoryUrl", "https://www.github.com/NuGet/NuGetGallery")]

```

File: tests/GitHubVulnerabilities2Db.Facts/Properties/AssemblyInfo.cs [hunk 20]
```diff
@@ -0,0 +1,36 @@
...
+
+// General Information about an assembly is controlled through the following
+// set of attributes. Change these attribute values to modify the information
+// associated with an assembly.
+[assembly: AssemblyTitle("GitHubVulnerabilities2Db.Facts")]
+[assembly: AssemblyDescription("")]
+[assembly: AssemblyConfiguration("")]
+[assembly: AssemblyCompany("")]
+[assembly: AssemblyProduct("GitHubVulnerabilities2Db.Facts")]
+[assembly: AssemblyCopyright("Copyright ©  2019")]
```
Comment (xavierdecoster) [inline tests/GitHubVulnerabilities2Db.Facts/Properties/AssemblyInfo.cs]:
```suggestion
[assembly: AssemblyCopyright("\x00a9 .NET Foundation. All rights reserved.")]
```

File: src/GitHubVulnerabilities2Db/Properties/AssemblyInfo.cs [hunk 21]
```diff
@@ -0,0 +1,36 @@
+﻿using System.Reflection;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+// General Information about an assembly is controlled through the following
+// set of attributes. Change these attribute values to modify the information
+// associated with an assembly.
+[assembly: AssemblyTitle("GitHubVulnerabilities2Db")]
+[assembly: AssemblyDescription("")]
```
Comment (xavierdecoster) [inline src/GitHubVulnerabilities2Db/Properties/AssemblyInfo.cs]:
```suggestion
[assembly: AssemblyDescription("Background job to ingest GitHub vulnerability data into Gallery database.")]
```

File: tests/GitHubVulnerabilities2Db.Facts/Properties/AssemblyInfo.cs [hunk 22]
```diff
@@ -0,0 +1,36 @@
...
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+// General Information about an assembly is controlled through the following
+// set of attributes. Change these attribute values to modify the information
+// associated with an assembly.
+[assembly: AssemblyTitle("GitHubVulnerabilities2Db.Facts")]
+[assembly: AssemblyDescription("")]
+[assembly: AssemblyConfiguration("")]
+[assembly: AssemblyCompany("")]
```
Comment (xavierdecoster) [inline tests/GitHubVulnerabilities2Db.Facts/Properties/AssemblyInfo.cs]:
```suggestion
[assembly: AssemblyCompany(".NET Foundation")]
```

File: tests/GitHubVulnerabilities2Db.Facts/Properties/AssemblyInfo.cs [hunk 23]
```diff
@@ -0,0 +1,36 @@
+﻿using System.Reflection;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+// General Information about an assembly is controlled through the following
+// set of attributes. Change these attribute values to modify the information
+// associated with an assembly.
+[assembly: AssemblyTitle("GitHubVulnerabilities2Db.Facts")]
+[assembly: AssemblyDescription("")]
+[assembly: AssemblyConfiguration("")]
```
Comment (xavierdecoster) [inline tests/GitHubVulnerabilities2Db.Facts/Properties/AssemblyInfo.cs]:
```suggestion
#if DEBUG
[assembly: AssemblyConfiguration("Debug")]
#else
[assembly: AssemblyConfiguration("Release")]
#endif
```

File: tests/GitHubVulnerabilities2Db.Facts/Properties/AssemblyInfo.cs [hunk 24]
```diff
@@ -0,0 +1,36 @@
+﻿using System.Reflection;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+// General Information about an assembly is controlled through the following
+// set of attributes. Change these attribute values to modify the information
+// associated with an assembly.
+[assembly: AssemblyTitle("GitHubVulnerabilities2Db.Facts")]
+[assembly: AssemblyDescription("")]
```
Comment (xavierdecoster) [inline tests/GitHubVulnerabilities2Db.Facts/Properties/AssemblyInfo.cs]:
```suggestion
[assembly: AssemblyDescription("Unit tests for the GitHubVulnerabilities2Db project.")]
```

File: tests/GitHubVulnerabilities2Db.Facts/Properties/AssemblyInfo.cs [hunk 25]
```diff
@@ -0,0 +1,36 @@
...
+//      Major Version
+//      Minor Version
+//      Build Number
+//      Revision
+//
+// You can specify all the values or you can default the Build and Revision Numbers
+// by using the '*' as shown below:
+// [assembly: AssemblyVersion("1.0.*")]
+[assembly: AssemblyVersion("1.0.0.0")]
+[assembly: AssemblyFileVersion("1.0.0.0")]
```
Comment (xavierdecoster) [inline tests/GitHubVulnerabilities2Db.Facts/Properties/AssemblyInfo.cs]: Think you can remove these lines. Won't the build automatically inject the version attributes?

Comment (xavierdecoster) [inline tests/GitHubVulnerabilities2Db.Facts/Properties/AssemblyInfo.cs]: Other projects in this solution have the following:

```suggestion
// The build will automatically inject the following attributes:
// AssemblyVersion, AssemblyFileVersion, AssemblyInformationalVersion, AssemblyMetadata (for Branch, CommitId, and BuildDateUtc)

[assembly: AssemblyMetadata("RepositoryUrl", "https://www.github.com/NuGet/NuGetGallery")]

```

File: src/GitHubVulnerabilities2Db/GitHubVulnerabilities2Db.csproj [hunk 26]
```diff
@@ -0,0 +1,115 @@
...
+    <None Include="Scripts\*" />
+  </ItemGroup>
+  <ItemGroup>
+    <PackageReference Include="MicroBuild.Core">
+      <Version>0.3.0</Version>
+      <IncludeAssets>runtime; build; native; contentfiles; analyzers</IncludeAssets>
+      <PrivateAssets>all</PrivateAssets>
+    </PackageReference>
+    <PackageReference Include="NuGet.Jobs.Common">
+      <Version>4.3.0-dev-3159955</Version>
```
[the remark this answers is not in the document]
Comment (scottbommarito) [reply src/GitHubVulnerabilities2Db/GitHubVulnerabilities2Db.csproj]: [I chose what was currently used in gallery master.](https://github.com/NuGet/NuGetGallery/blob/master/src/AccountDeleter/AccountDeleter.csproj#L110)

I think it's fine pointing to dev branch, but we probably should eventually align with a stable.

File: src/GitHubVulnerabilities2Db/Properties/AssemblyInfo.cs (hunk 19 above)
Comment (scottbommarito) [reply src/GitHubVulnerabilities2Db/Properties/AssemblyInfo.cs]: I don't think so as well--I didn't put much attention into this file, just kept what it autogenerated.

File: src/GitHubVulnerabilities2Db/Properties/AssemblyInfo.cs (hunk 21 above)
Comment (scottbommarito) [reply src/GitHubVulnerabilities2Db/Properties/AssemblyInfo.cs]: assembly description

File: src/GitHubVulnerabilities2Db/GraphQL/SecurityAdvisory.cs [hunk 27]
```diff
@@ -0,0 +1,49 @@
...
+
+using System;
+using System.Collections.Generic;
+
+namespace GitHubVulnerabilities2Db.GraphQL
+{
+    /// <summary>
+    /// https://developer.github.com/v4/object/securityadvisory/
+    /// </summary>
+    public class SecurityAdvisory : INode
```
[the remark this answers is not in the document]
Comment (scottbommarito) [reply src/GitHubVulnerabilities2Db/GraphQL/SecurityAdvisory.cs]: moved `SecurityVulnerability` and `SecurityVulnerabilityPackage` to their own file--I don't think it's worth giving the tiny classes their own file

File: src/GitHubVulnerabilities2Db/GraphQL/QueryResponse.cs (hunk 14 above)
Comment (scottbommarito) [reply src/GitHubVulnerabilities2Db/GraphQL/QueryResponse.cs]: moved `INode` and `Edge` into their own class, but I think `QueryResponse`, `QueryResponseData`, and `ConnectionResponseData` are similar and short enough to be together.

File: src/GitHubVulnerabilities2Db/GraphQL/QueryService.cs (hunk 15 above)
Comment (scottbommarito) [reply src/GitHubVulnerabilities2Db/GraphQL/QueryService.cs]: I didn't add tests to the `QueryService` because I felt the logic was quite simple, but I might as well haha.

File: src/GitHubVulnerabilities2Db/Collector/AdvisoryQueryService.cs [hunk 28]
```diff
@@ -0,0 +1,115 @@
...
+        {
+            // We want to keep the next advisory's data, but prepend the existing vulnerabilities that were returned in previous queries.
+            nextAdvisory.Vulnerabilities.Edges = advisory.Vulnerabilities.Edges.Concat(
+                nextAdvisory.Vulnerabilities.Edges ?? Enumerable.Empty<Edge<SecurityVulnerability>>());
+            // We are not querying the advisories feed at this time so we do not want to advance the advisory cursor past what it was originally.
+            nextAdvisory.UpdatedAt = advisory.UpdatedAt;
+            return nextAdvisory;
+        }
+
+        private class VulnerabilityForSameAdvisoryComparer : IEqualityComparer<SecurityVulnerability>, IEqualityComparer<Edge<SecurityVulnerability>>
```
[the remark this answers is not in the document]
Comment (scottbommarito) [reply src/GitHubVulnerabilities2Db/Collector/AdvisoryQueryService.cs]: `AdvisoryQueryServiceFacts` already tests that it doesn't find different `SecurityVulnerabilities` to be the same. I think I can add a test to `AdvisoryQueryServiceFacts` that expects that it dedupes and I think that's sufficient.

File: src/GitHubVulnerabilities2Db/Collector/AdvisoryQueryService.cs [hunk 29]
```diff
@@ -0,0 +1,115 @@
...
+                var nextQuery = _queryBuilder.CreateSecurityAdvisoriesQuery(afterCursor: lastCursor);
+                var nextResponse = await _queryService.QueryAsync(nextQuery, token);
+                lastAdvisoryEdges = nextResponse?.Data?.SecurityAdvisories?.Edges?.ToList() ?? Enumerable.Empty<Edge<SecurityAdvisory>>();
+                advisories.AddRange(lastAdvisoryEdges.Select(e => e.Node));
+            }
+
+            return await Task.WhenAll(advisories.Select(a => FetchAllVulnerabilities(a, token)));
+        }
+
+        private async Task<SecurityAdvisory> FetchAllVulnerabilities(SecurityAdvisory advisory, CancellationToken token)
```
Comment (loic-sharma) [inline src/GitHubVulnerabilities2Db/Collector/AdvisoryQueryService.cs]:
```suggestion
        private async Task<SecurityAdvisory> FetchAllVulnerabilitiesAsync(SecurityAdvisory advisory, CancellationToken token)
```

Comment (scottbommarito) [reply src/GitHubVulnerabilities2Db/Collector/AdvisoryQueryService.cs]: done!

File: src/GitHubVulnerabilities2Db/Ingest/GitHubVersionRangeParser.cs [hunk 30]
```diff
@@ -0,0 +1,104 @@
...
+            // Remove commas in version range. They exist solely for readability.
+            gitHubVersionRange = gitHubVersionRange.Replace(",", string.Empty);
+
+            // A GitHub version range consists of pairs of:
+            // 1. A symbol (<, >, <=, or >=), which defines whether the next version is the minimum or maximum and whether or not it's included or excluded in the range.
+            // 2. A SemVer version.
+            var versionRangeParts = gitHubVersionRange.Split(' ');
+            if (versionRangeParts.Length > 4)
+            {
+                throw new ArgumentException("A version range cannot contain more than two pairs.");
```
Comment (loic-sharma) [inline src/GitHubVulnerabilities2Db/Ingest/GitHubVersionRangeParser.cs]: If these exceptions ever happen we will need to fix them. I would include the raw `gitHubVersionRange` input to all these exception message.

Comment (scottbommarito) [reply src/GitHubVulnerabilities2Db/Ingest/GitHubVersionRangeParser.cs]: Good call! It would be hard to debug this otherwise.

File: src/GitHubVulnerabilities2Db/Ingest/IAdvisoryIngestor.cs [hunk 31]
```diff
@@ -0,0 +1,17 @@
...
+using GitHubVulnerabilities2Db.GraphQL;
+
+namespace GitHubVulnerabilities2Db.Ingest
+{
+    /// <summary>
+    /// Processes new or updated <see cref="SecurityAdvisory"/>s.
+    /// </summary>
+    public interface IAdvisoryIngestor
+    {
+        Task Ingest(IReadOnlyList<SecurityAdvisory> advisories);
```
Comment (loic-sharma) [inline src/GitHubVulnerabilities2Db/Ingest/IAdvisoryIngestor.cs]:
```suggestion
        Task IngestAsync(IReadOnlyList<SecurityAdvisory> advisories);
```
